IPLD-Data-Model-v1.md

```diff
@@ -1,17 +1,108 @@
-# [WIP] IPLD Data Model
+IPLD Data Model
+===============
 
-## Simple Types
+The IPLD Data Model is a core part of the IPLD specification,
+which describes what data is representable in IPLD --
+for example, booleans, integers, textual strings, maps and lists, etc.
+
+While the Data Model describes these representations in the abstract,
+[Codecs](Codecs) specify exactly how these data are transcribed into serialized bytes.
+(Another component of the IPLD specifications, [Schemas](schema), provide
+additional optional tooling on top of the Data Model which can further refine,
+describe, and constrain the range of acceptable data values.)
+
+
+Kinds
+-----
+
+The following is the list of essential _kinds_ (or more formally, _representation kinds_)
+of data representable in the IPLD Data Model:
 
-* Boolean
 * Null
+* Boolean
+* Integer
+* Float
 * String
+* Bytes
+* List
+* Map
+* Link
+
+(Note that we use the term "_kinds_" here to disambiguate this from "_types_",
+which is a term we'll use at the [Schemas](schema) level.)
+
+The _recursive kinds_ are:
+
+* List
+* Map
+
+The _scalar kinds_ (the complement of recursive) are:
+
+* Null
+* Boolean
 * Integer
 * Float
-* Array
-* Object (Hash Map)
-* Binary
+* String
+* Bytes
+* Link
+
+(Note that [Schemas](schema) introduce a few more kinds -- when clarification is necessary,
+these Data Model kinds can be called the "_representation kinds_",
+while the additional kinds introduced in the Schema layer are "_perceived kinds_".)
+
+### Kinds Reference
+
+Each of the following sections provides details about each of the kinds
+introduced in the summary table above.
+
+#### Null kind
+
+Null is a scalar kind.  Its cardinality is one -- the only value is 'null'.
 
-## Link Type
+#### Boolean kind
 
-This type represents a link to another IPLD Block. The link reference
+Boolean is a scalar kind.  Its cardinality is two -- either the value 'true' or the value 'false'.
+
+#### Integer kind
+
+
+#### Float kind
+
+
+#### String kind
+
+String is a scalar kind.  Its cardinality is infinite -- strings do not have a length limit.
+
+Strings should be UTF-8 text, in [NFC](https://www.unicode.org/reports/tr15/#Norm_Forms) canonicalization.
+
+#### Bytes kind
+
+Bytes is a scalar kind.  Its cardinality is infinite -- byte sequences do not have a length limit.
+
+Bytes are distinct from strings in that they are not considered to have any character encoding nor
+generally expected to be printable as human-readable text.
+In order to print byte sequences as text, additional effort such as Base64 encoding may be required.
+
+#### List kind
+
+List is a recursive kind.
+
+Values contained in lists can be accessed by their ordinal offset in the list.
+
+#### Map kind
+
+Map is a recursive kind.
+
+Values in maps are accessed by their "key".  Maps can also be iterated over,
```

Comment:
do we have a restriction on the kind(s) that can be used for Map keys? just strings? strings and integers? arbitrary?

Reply:
Casually, there seems to be a consensus we're going to go with strings.  The ecosystemic costs of having non-printable map keys (e.g. bytes) seem to be... High.

Integer map keys can sometimes be useful, but anecdotally, the most common response on hearing they're disallowed is "harumph.  okay." rather than any more ardent objections; while on the costs side, integer map keys are fraught with peril and nonsense in practice in languages like javascript, and they're possible but add nontrivial implementation complexity in go...  There'd be benefits to supporting integer keys, no doubt, but the cost/benefit ratio seems rough.

On a bright side, one cool thing we can do with the schema-vs-representation-kinds relationship: supposing we spec that map keys must be of _representation kind_ string... we can in fact still allow complex keys, even structs, as map keys.  They just need a representation strategy like e.g. 'stringjoin'.

Reply:
Personally, I'd _kind_ of like to have bytes, using multibase in paths. I believe this should "just work" and should improve some datastructures quite a bit.

Reply:
Most formats don’t support non-string keys so it would be quite the ask to add this to the data model layer. 

Adding it to the Map type over in schema land is more doable, but I’d still like to have a base Map type that only supports strings and we can define some fancier Map type with support for bytes as keys.

Reply:
Sounds fine. Base Map key type is strings only. Collections built on top can optionally have more sophisticated types.

So far in my collections experiments [I've just been targeting byte arrays](https://github.com/rvagg/iamap#IAMap_set) (or things that can be easily converted to them) because they are the easiest thing to hash and give you maximal flexibility. I can imagine some folks wanting specific types or cleaner interfaces for other types (i.e. without having to jump through the convert-to-byte-array hoops) and I think I've opened the door for that possibility in https://github.com/ipld/specs/pull/110, but it'll likely be an exercise for the people who want that.